core/vdbe/execute.rs

```diff
@@ -1560,41 +1564,70 @@ pub fn op_column(
-                    n if n >= 13 && n % 2 == 1 => Value::Text(Text {
-                        value: data_slice.to_vec(),
-                        subtype: TextSubtype::Text,
-                    }),
-                    _ => default.clone().unwrap_or(Value::Null),
+                    n if n >= 12 && n % 2 == 0 => {
+                        Some(RefValue::Blob(RawSlice::create_from(data_slice)))
+                    }
+                    n if n >= 13 && n % 2 == 1 => Some(RefValue::Text(TextRef::create_from(
+                        data_slice,
+                        TextSubtype::Text,
+                    ))),
+                    _ => None,
                 }
             };
 
+            let Some(value) = value else {
+                // DEFAULT handling. Try to reuse the registers when allocation is not needed.
+                let Some(ref default) = default else {
+                    state.registers[*dest] = Register::Value(Value::Null);
+                    state.pc += 1;
+                    return Ok(InsnFunctionStepResult::Step);
+                };
+                match (default, &mut state.registers[*dest]) {
+                    (Value::Text(new_text), Register::Value(Value::Text(existing_text))) => {
+                        if !existing_text.maybe_extend(new_text) {
+                            state.registers[*dest] = Register::Value(default.clone());
+                        }
+                    }
+                    (Value::Blob(new_blob), Register::Value(Value::Blob(existing_blob))) => {
+                        if !existing_blob.maybe_extend(new_blob) {
+                            state.registers[*dest] = Register::Value(default.clone());
+                        }
+                    }
+                    _ => {
+                        state.registers[*dest] = Register::Value(default.clone());
+                    }
+                }
+                state.pc += 1;
+                return Ok(InsnFunctionStepResult::Step);
+            };
+
             // Try to reuse the registers when allocation is not needed.
             match (&value, &mut state.registers[*dest]) {
-                (Value::Text(new_text), Register::Value(Value::Text(existing_text))) => {
-                    if existing_text.value.capacity() >= new_text.value.len() {
-                        existing_text.value.clear();
-                        existing_text.value.extend_from_slice(&new_text.value);
-                        existing_text.subtype = new_text.subtype;
-                    } else {
-                        state.registers[*dest] = Register::Value(value);
+                (RefValue::Text(new_text), Register::Value(Value::Text(existing_text))) => {
+                    if !existing_text.maybe_extend(new_text) {
+                        state.registers[*dest] =
+                            Register::Value(Value::Text(Text::new(new_text.as_str())));
```

Comment:
Why not extend always? If there is a register with allocated memory, we should just try to extend the buffer as much to fit this new text instead of allocating a new value.

Reply:
True. This was the existing implementation but it doesn't really make sense not to extend every time

Reply:
implemented in https://github.com/tursodatabase/turso/pull/2205/commits/111c0032ae8c7e29d6bd235a35c9096137f0a917